custom_components/localtuya/fan.py

```diff
@@ -74,38 +92,60 @@ async def async_turn_off(self, **kwargs) -> None:
 
     async def async_set_speed(self, speed: str) -> None:
         """Set the speed of the fan."""
+        self._speed = speed
+
         if speed == SPEED_OFF:
             await self._device.set_dps(False, "1")
-        elif speed == SPEED_LOW:
-            await self._device.set_dps("1", "2")
-        elif speed == SPEED_MEDIUM:
-            await self._device.set_dps("2", "2")
-        elif speed == SPEED_HIGH:
-            await self._device.set_dps("3", "2")
+        else:
+            await self._device.set_dps(
+                self._config.get(speed), self._config.get(CONF_FAN_SPEED_CONTROL)
+            )
+
         self.schedule_update_ha_state()
 
     async def async_oscillate(self, oscillating: bool) -> None:
         """Set oscillation."""
-        await self._device.set_dps(oscillating, "8")
+        await self._device.set_dps(
+            oscillating, self._config.get(CONF_FAN_OSCILLATING_CONTROL)
+        )
         self.schedule_update_ha_state()
 
     @property
     def supported_features(self) -> int:
         """Flag supported features."""
-        return SUPPORT_SET_SPEED | SUPPORT_OSCILLATE
+        if self.has_config(CONF_FAN_OSCILLATING_CONTROL):
```

Comment:
Change to this pattern instead:

https://github.com/rospogrigio/localtuya-homeassistant/blob/6693d9061575d1b0c45e2a8cc7e87bbb4f316515/custom_components/localtuya/light.py#L106

Speed is optional (according to `flow_schema`) so we might as well handle that case as well.

Reply:
Do we really want speed being optional?

Reply:
I've been thinking about that too. Maybe if there are any really basic fans that doesn't support different speeds?